fix(compiler): Don't strip `/*# ... */` comment when css scoping

asnowwolf opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x")
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)

Currently, the source-mapping comments of component style is stripped by extractStyleUrls, so, css mappings don't work properly for component styles.

**What is the new behavior?**

Don't strip hash comments, so that they will be past to `style` tags properly.

**Does this PR introduce a breaking change?** (check one with "x")
- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...


**Other information**:

asnowwolf commented:
Sorry, when i merge the latest branch, a misuse destroys #16090 , this one is a re-pushed version that has merged the latest master.

vicb commented on packages/compiler/src/style_url_resolver.ts:
```diff
@@ -29,19 +29,20 @@ export function extractStyleUrls(
-        foundUrls.push(resolver.resolve(baseUrl, url));
-        return '';
-      });
+  const modifiedCssText = cssText.replace(CSS_COMMENT_WITH_NO_SOURCE_MAP_REGEXP, '')
```
could we make it more robust by using a replace fn that would:
- check if `[0] === '#'`
- check if the content contains `sourceURL=` or `sourceMappingURL=`

asnowwolf replied:
Did you mean that all of the comments that begin with # are likely to leave something that we do not want to keep, So we should only ignore the known sourceURL and sourceMappingURL? If so, I will try to revise it.

asnowwolf replied:
There are some questions I can not decide: whether we should support the legacy format `/*  @`, variant format `// #` and abnormal format `/*#` (no spaces between `*` and `#`, it's illegal, but exists.)

Commit:
fix(compiler): Don't strip CSS source maps

Fix CSS source mapping for component by keeping `/*# sourceMappingURL= ... */` and  `/*# sourceURL= ... */` comments.

Relates to <https://github.com/angular/angular-cli/issues/4199>.

asnowwolf commented:
There are some questions: whether we should support the legacy format `/*  @`, and abnormal format `/*#` (no spaces between `*` and `#`, it's illegal, but exists.)

asnowwolf commented:
It seems that CI has failed, but I have been tested in the local development environment, and CI's failure seems to be due to random factors. Do I need to trigger CI until it runs successfully?

Toxicable commented:
Looks due to a flake to me, i'd give suggest giving it a rerun

asnowwolf commented:
The `aio-e2e-*` tasks failed. I think this is the problem of CI itself.

gkalpak commented on packages/compiler/src/style_url_resolver.ts:
```diff
@@ -29,19 +29,31 @@ export function extractStyleUrls(
 
 const CSS_IMPORT_REGEXP = /@import\s+(?:url\()?\s*(?:(?:['"]([^'"]*))|([^;\)\s]*))[^;]*;?/g;
-const CSS_COMMENT_REGEXP = /\/\*[\s\S]+?\*\//g;
+const CSS_COMMENT_REGEXP = /\/\*([\s\S]+?)\*\//g;
```
I think you could just change the RegExp to `/\/\*(?!# source(?:Mapping)?URL=)[\s\S]+?\*\//g` and leave everything else as it was.
(Not that it matters much :grin:)

asnowwolf replied:
Yes, I think it's better too.

vicb commented:
The changes LGTM.

We should merge to master only as the behavior is different (we now keep comment when they contain source maps).

@asnowwolf could you please change the commit message to match the changes you've made and give more info:

```
fix(compiler): Don't strip CSS source maps 

Fix CSS source mapping for component by keeping `/*# sourceMappingURL= ... */` and  `/*# sourceURL= ... */` comments
```

Also please add links to the issues related to this PR (`relates to #....` of `fixes #...`).

asnowwolf commented:
Done, but here is not a related issue, the relevant issue in the angular-cli repository.

vicb commented:
Changes LGTM.
Travis was red, restarted.

asnowwolf commented:
BTW:

What is going to do with #16088 ? It is aimed at a BUG that appears only in Chrome (and possibly a feature in Chrome because it's been there for a long time, so there's a special Chrome plugin to fix it). In addition, it can be modified by modifying the Angular-CLI or other toolchain.

Is it worth merging here? Do I need to turn it off and to find other ways to solve it?

mwijnands commented:
Can this be fixed in a 4.x release as well? Behavior changed but how would it affect anything? Also it's mentioned in the changelog as a bugfix, not a breaking change.
